helper/schema/schema.go

```diff
@@ -1414,6 +1414,21 @@ func (m schemaMap) validate(
 			"%q: this field cannot be set", k)}
 	}
 
+	err := validateRequiredWithAttribute(k, schema, c)
+	if err != nil {
+		return nil, []error{err}
+	}
+
+	err = validateExactlyOneAttribute(k, schema, c)
```

Comment:
Actually that's why I initially filed the PR: before the discussion here, it made no sense to me checking anything on an optional property which has no value. That's why my explanation in the chapter _Semantic changes in validation for optional schema fields_ above. 

After yours and @appilon's explanation how you guys want to use the various validation "operators" I fully understand why you come up with the request to change the implementation. 

But the question remains: how can I define validation inside the schema that a property needs other properties **but only if it has a value**? If you always validate the property regardless it it has a value or not, how do I implement my scenario?

Reply:
I noted in the very last sentence of my last comment that you're right that RequiredWith needs to happen right where it is. This is especially true because of the reason you just stated

Does RequiredWith hit all the checks you need now or are we still missing some functionality that warrants another validator?

Reply:
I was aware what you said about the position of `RequiredWith` but my question included also the two other validation functions (`validateExactlyOneAttribute` and `validateAtLeastOneAttribute`). 

If you leave them at the old position, form my understanding, there's no possibility to model the requirement, that a validation is only performed on an optional attribute when it has a value (or a default value). 

I know that this collides with your statements about how you wanna use the `AtLeastOne` attribute to ensure that at least one of the a set of properties has been set for a set of optional attributes.

So, my question, how to model the requirement, that a validation is only performed when a property has a value? Is this even possible in the current configuration? Do miss something? Especially when I take into account the other options in the schema, like `Optional`, `Required` etc.

Reply:
I need a little more confirmation on what exactly you are looking for. Are you looking for ExactlyOneOf/AtLeastOneOf validation to be only be run when a value has been set? Or are where these checks preventing any form of validation from being run? 

If you're looking for ExactlyOneOf functionality for only when an attribute has been set than you should just use ConflictsWith. That ensures that only one value in that set of keys can be set at a time and only is checked when the key in question is set. 

AtLeastOneOf validation is a little trickier since its goal is to have at least one attribute in a list of keys to be set. If you only want that validation for when a key is set then the whole point is moot because the key is just Optional at that point. 

Do you have a real world example you can share that would better help me see what you're trying to achieve? 

Regardless, making changes to ExactlyOneOf/AtLeastOneOf is outside the scope of adding RequiredWith. We should move them back and open an issue on missing functionality so we can track that in a single issue and move forward on getting this piece in since many developers (myself included) could make good use of RequiredWith today.

Reply:
To give you a real world example I can share the configuration for setting Git permissions via the [Azure DevOps Terraform provider](https://github.com/microsoft/terraform-provider-azuredevops):

The schema definition looks like the following:
```go
func resourceGitPermissions() *schema.Resource {
	return &schema.Resource{
		Create: resourceGitPermissionsCreate,
		Read:   resourceGitPermissionsRead,
		Update: resourceGitPermissionsUpdate,
		Delete: resourceGitPermissionsDelete,
		Importer: &schema.ResourceImporter{
			State: resourceGitPermissionsImporter,
		},
		Schema: securityhelper.CreatePermissionResourceSchema(map[string]*schema.Schema{
			"project_id": {
				Type:         schema.TypeString,
				ValidateFunc: validate.UUID,
				Required:     true,
				ForceNew:     true,
			},
			"repository_id": {
				Type:         schema.TypeString,
				ValidateFunc: validate.UUID,
				Optional:     true,
				ForceNew:     true,
			},
			"branch_name": {
				Type:         schema.TypeString,
				ValidateFunc: validate.NoEmptyStrings,
				Optional:     true,
				ForceNew:     true,
				ExactlyOneOf: []string{"repository_id"},
			},
		}),
	}
}
```

As you can see the property `branch_name` is optional with no default value. As the same with the `repository_id`. But if you wanna set some permissions on a Git branch, you always have to specify the  `repository_id` because otherwise the Azure DevOps REST Api doesn't know where the branch is located.

If I use the following HCL code to set the permissions for various locations inside the Git projects hierarchy, validation will always fail, because `ExactlyOneOf` is validated even if the (optional) property `branch_name` isn't used.

```hcl
resource "azuredevops_git_permissions" "tf-project-test-git-root-permissions" {
  project_id  = local.project_map[var.project_name].project_id  
  principal   = data.azuredevops_group.tf-project-contributors.id
  permissions = {
    CreateRepository = "Deny"
    DeleteRepository = "Deny"
    RenameRepository = "NotSet"
  }
}

resource "azuredevops_git_permissions" "tf-project-test-git-repo-permissions" {
  project_id    = local.project_map[var.project_name].project_id
  repository_id = var.git_repository_id
  principal     = data.azuredevops_group.tf-project-administrators.id
  permissions   = {
    RemoveOthersLocks = "Allow"
    ManagePermissions = "Deny"
    CreateTag         = "Deny"
    CreateBranch      = "NotSet"
  }
}

resource "azuredevops_git_permissions" "tf-project-test-git-branch-permissions" {
  project_id    = local.project_map[var.project_name].project_id
  repository_id = var.git_repository_id
  branch_name   = var.git_repository_branch
  principal     = data.azuredevops_group.tf-project-build-administrators.id
  permissions   = {
    RemoveOthersLocks = "Allow"
    ForcePush         = "Deny"
  }
}
```

![image](https://user-images.githubusercontent.com/14177833/79046188-ab638080-7c0f-11ea-952a-d0b4b7c59643.png)

I am very surprised that this issue has never surfaced before. That's why I moved the checks for `ExactlyOneOf` and `AtLeastOneOf` to their new positions. But I agree with you that this issue should discussed in a separate issue (pull request) so that people can quickly benefit from `RequiredWith` and if we leave the `RequiredWith` at the position where it located now, I do have a workaround in the Azure DevOps provider, because instead of `ExactlyOneOf` I can utilize `RequiredWith`. 

Semantically this is a horror because `ExactlyOneOf` and `AtLeastOneOf` are validated differently by the SDK than `RequiredWith`, but we should be pragmatic here. Therefore I will move the tests for `ExactlyOneOf` and `AtLeastOneOf` back to the original location and update the documentation of the pull request.

Reply:
Ah, so this isn't a bug but an issue with documentation about how these validators are supposed to be used. 

ExactlyOneOf and AtLeastOneOf use all keys including the calling key (branch_name in your case) to confirm that only one of those keys is set. So `ExactlyOneOf: []string{"repository_id"},` is in fact `ExactlyOneOf: []string{"repository_id", "branch_name"},` and we're confirming that one and only one of `repository_id` or `branch_name` are specified. 

As we designed these validators, we knew that a user might do what you've done and not specify the calling key in their list and so we made sure to add it to the list so we could get proper validation. But we didn't make it well known that this was what was happening. I'll update the documentation so this information is readily available.

Reply:
@mbfrahry Okay, so my use case _"optional property, which requires another property to be set as well when specified itself"_ can't be implemented using `ExactlyOneOf` and `AtLeastOneOf`. Right? 

I don't think my use case is an edge case. So what's the (planned) solution from the SDK team?

Reply:
What your wrote with RequiredWith solves this, right? When one is set, the other must be set. 

We'll get this feature out shortly. Doing the last bit of testing now

Reply:
Yes the `RequiredWith` will solve it for now. But to be clear, I'm not very comfortable with such a "design" where validations functions have subtle (not obvious, surprising) differences how they are evaluated. Especially what if, if I wanna model a dependency for an optional property that requires exactly or at least ONE additional other property to be specified from a list of other properties. Not feasible right now if I understood or discussion completely right. Not that ideal from my point of view. But perhaps I'm overemphasizing the validation stuff, because in the end, a creator of a provider can issue an error in the plan or in the apply phase..

Reply:
For the use case you describe you should use a TypeList/TypeSet to set up the optional property. In that, we can have use AtLeastOneOf or ExactlyOneOf on the attributes in that List/Set. This models a dependency where the attribute is optional and you require exactly or at least one additional other property to be specified. 

ExactlyOneOf/AtLeastOneOf was written for a very specific purpose where `Required/Optional` tags were not enough and extra logic was written into the provider that was only executed at apply time. I saw this as an opportunity to have the sdk do that validation for me at plan time. I apologize for missing the mark on documenting how to use  these validators and will hopefully fix that shortly.

Reply:
Referencing my sample above and according to your last explanation, I should model the  `branch_name`property using a List/Set in which I include a (sub-)schema which contains the other (optional) attributes (I my case only one) and in this sub-schema I can use the ExactlyOneOf/AtLeastOneOf attributes to ensure that at least of them is specified?

Reply:
For your `branch_name/repository_id` example, the current best practice is the following:
```
"github_configuration": {
  Type: schema.TypeList
  MaxItems: 1,
  Optional: true,
  Elem: &schema.Resource {
    Schema: map[string]*schema.Schema{
      "repository_id": {
        Required: true,
        Type: schema.TypeString,
      },
      "branch_name": {
        Required: true,
        Type: schema.TypeString,
      },
    },
  },
},
```

This format allows you to show exactly what attributes are needed to configure with github properly while also leaving you open to add other vcs configuration attributes. I'd recommend you go down this path rather than using `RequiredWith` for the ease on the users (in knowing what attributes go to which vcs attributes) and maintainers (in case you add more vcs options).

Reply:
Sorry but this makes no sense to me, because I've to model something in the SDK that isn't there in the semantics how you use the backend REST API, in that case the Azure DevOps API. 

As you can see in the above sample (the HCL code, obviously) it's perfectly valid to have none of `repository_id` and `branch_name` specified because in that case you change the permissions for all Git repositories in Azure DevOps for a specific project.

The next hierarchy level down is the repository. So to specify permissions that get inherited down to all branches inside a Git repository you have to specify the `project_id` and the `repository_id` . Because the `project_id` is required anyway, we have no issues here during validation.

The last hierarchy level down is the branch So to specify permissions on a specific branch inside a Git repository you (obviously) have to specify the `project_id`, the `repository_id` and the `branch_name`. But because `repository_id` and `branch_name` are both optional, the user might forget to specify the `repository_id` while setting the `branch_name` attribute. Because how validation is implemented right now, this can't be discovered during the validation phase and while throw an error during plan or apply. What's definitly is okay but to be user friendly it should be possible to show errors in the HCL code as soon as possible.

So according to your proposal the I'd get probably away with some kind of HCL code like the following:

```hcl
resource "azuredevops_git_permissions" "tf-project-test-git-root-permissions" {
  project_id  = local.project_map[var.project_name].project_id  
  principal   = data.azuredevops_group.tf-project-contributors.id
  permissions = {
    CreateRepository = "Deny"
    DeleteRepository = "Deny"
    RenameRepository = "NotSet"
  }
}

resource "azuredevops_git_permissions" "tf-project-test-git-repo-permissions" {
  project_id    = local.project_map[var.project_name].project_id  
  repository_configuration {
      repository_id = var.git_repository_id
  }
  principal     = data.azuredevops_group.tf-project-administrators.id
  permissions   = {
    RemoveOthersLocks = "Allow"
    ManagePermissions = "Deny"
    CreateTag         = "Deny"
    CreateBranch      = "NotSet"
  }
}

resource "azuredevops_git_permissions" "tf-project-test-git-branch-permissions" {
  project_id    = local.project_map[var.project_name].project_id
  branch_configuration {
    repository_id = var.git_repository_id
    branch_name   = var.git_repository_branch
  }
  principal     = data.azuredevops_group.tf-project-build-administrators.id
  permissions   = {
    RemoveOthersLocks = "Allow"
    ForcePush         = "Deny"
  }
}
```

To me hard to understand and sorry only enforced by the quirks inside the SDK validation.

Reply:
So my proposal would be to add validation functions like `AtLeastOneOf` or `ExactlyOneOf` (or both, or with other names) to the `Resource` and move the tests for `AtLeastOneOf` or `ExactlyOneOf` for schema properties after the test if a value is specified.

Doing that will give you the ability to check, if the user specified all required properties for a resource (that's the use case you came up with, deprecation of a property) and on the property level you have the ability to enforce other type of requirements that them from semantics around the property in question.

To me, much more concise and easier to use and understand by the user of the SDK.

Reply:
> Sorry but this makes no sense to me, because I've to model something in the SDK that isn't there in the semantics how you use the backend REST API, in that case the Azure DevOps API.
I believe that a core tenant of Terraform is to make configuring infrastructure easy and simple from a users perspective. APIs can be notoriously tricky and complicated. We try to match how a user would want to configure a resource rather than mirror 1:1 what an API wants. 

From your comments, `repository_id` is required when specifying anything github related. So we can change my solution to instead look like
```
"project_id": {
  Required: true,
  Type: schema.TypeString,
},
"github_configuration": {
  Type: schema.TypeList
  MaxItems: 1,
  Optional: true,
  Elem: &schema.Resource {
    Schema: map[string]*schema.Schema{
      "repository_id": {
        Required: true,
        Type: schema.TypeString,
      },
      "branch_name": {
        Optional: true,
        Type: schema.TypeString,
      },
    },
  },
},
```

The above should work for the problem you're describing.  `project_id` must be specified at all times. `github_configuration` is optional. When `github_configuration` is specified, you must specify `repository_id`. `branch_name` is optional when specifying `github_configuration`. 

I apologize if I'm not seeing how all these pieces fit together but I have not looked into the azure devops api and am doing a best effort solution on the information you're providing me.

Reply:
> So my proposal would be to add validation functions like AtLeastOneOf or ExactlyOneOf (or both, or with other names) to the Resource and move the tests for AtLeastOneOf or ExactlyOneOf for schema properties after the test if a value is specified.

Moving all the validators to the Resource level has been on my radar for quite some time actually but time has not been on my side and from the brief time I looked into it, it's non-trivial.

Reply:
> I apologize if I'm not seeing how all these pieces fit together but I have not looked into the azure devops api and am doing a best effort solution on the information you're providing me.

If I sounded harsh I apologize as well. It wasn't definitely my intention to disqualify your efforts to bring up alternate solutions. On the contrary, they gave me a deeper insight into how the SDK works. Thanks for the extensive discussion! Highly appreciated.

Reply:
Unfortunately, much of provider development is still tribal knowledge and learning by doing. I wish this wasn't the case but we do have plans in the future to ease the learning curve. 

Happy to discuss further and you can @ me on a PR if you'd like me to review what solution you ultimately end up going with.

Reply:
Well I guess I'll use the `RequiredWith` constrained for the current implementation of the DevOps provider. Cheapest way to go and, according to HCL, the easiest to understand for the user of the provider.  When do you think the PR will be merged?

Aside I'm happy to know, that I can @ you on the DevOps PR or some other people here when I might file another PR for the SDK or if I've an issue with it. :+1: :+1:  Always hard to figure out who's in charge for a project on GitHub.

Reply:
I agree that that is the way you should go with `RequiredWith`! 

It's on the sdk team now to get it merged and released but they know it's ready to go!